sharedbuffers/mapped_struct.py

```diff
@@ -3869,6 +4049,447 @@ def merge(cls, parts, destfile = None, tempdir = None,
+                                pindex += stride0
+                        finally:
+                            PyBuffer_Release(cython.address(indexbuf))
+                    else:
+                        index = self.index
+                        while startpos < nitems and index[startpos,0] == hkey:
+                            if self._compare_keys(self._buf, self.index[startpos,1], key):
+                                return index[startpos,2]
+                            startpos += 1
+                finally:
+                    PyBuffer_Release(cython.address(pybuf))
+                #lint:enable
+            else:
+                index = self.index
+                while startpos < nitems and index[startpos,0] == hkey:
+                    if self._compare_keys(buf, index[startpos, 1], key):
+                        return index[startpos,2]
+                    startpos += 1
+        return default
+
+    @classmethod
+    @cython.locals(
+        basepos = cython.ulonglong, curpos = cython.ulonglong, endpos = cython.ulonglong, finalpos = cython.ulonglong,
+        dtypemax = cython.ulonglong)
+    def build(cls, initializer, destfile = None, tempdir = None, map_file=True):
+        if destfile is None:
+            destfile = tempfile.NamedTemporaryFile(dir = tempdir)
+
+        dtype = cls.dtype
+        try:
+            dtypemax = ~dtype(0)
+        except:
+            try:
+                dtypemax = ~dtype.type(0)
+            except:
+                dtypemax = ~0
+
+        basepos = destfile.tell()
+
+        # Reserve space for the header
+        write = destfile.write
+        write(cls._Header.pack(0, 0))
+
+        # Build the index - the index is a matrix of the form:
+        # [ [ hash, key offset, value id ], ... ]
+        #
+        # With the rows ordered by hash
+        if isinstance(initializer, dict):
+            initializer = initializer.iteritems()
+        else:
+            initializer = iter(initializer)
+        parts = []
+        islice = itertools.islice
+        array = numpy.array
+        curpos = basepos + cls._Header.size
+        pack_into = mapped_object.pack_into
+        valbuf = bytearray(65536)
+        valbuflen = 65536
+        n = 0
+        part = []
+        while 1:
+            del part[:]
+            for k,i in islice(initializer, 1000):
+                # Add the index item
+                n += 1
+                # Not *quite* as correct as computing the actual length in
+                # bytes, but since this is only an upper bound, we should be fine.
+                klen = sys.getsizeof(k)
+                if curpos > dtypemax:
+                    raise ValueError("Cannot represent offset with requested precision")
+                part.append((_stable_hash(k), curpos, i))
+                if klen + 16 > valbuflen:
+                    valbuflen = (klen + 16) * 2
+                    valbuf = bytearray(valbuflen)
+                endpos = pack_into(k, valbuf, 0)
```

Comment:
Why do you use a temp bytearray instead of simply packing into the destination buffer?

Reply:
I'm using another IdMapper as a template, and I guess the reasoning there is that using a temparray is easier than writing directly to a file.

Reply:
Sorry, yes, I thought build was building the idmapper into a buffer, but you're right, it's using a file.